monai/transforms/compose.py

```diff
@@ -358,3 +359,61 @@ def inverse(self, data):
         for o in reversed(applied_order):
             data = apply_transform(self.transforms[o].inverse, data, self.map_items, self.unpack_items, self.log_stats)
         return data
+
+
+class SomeOf(Compose):
+    """
+    ``SomeOf`` provides the ability to randomly sample a fixed number or
+    up to a fixed number of transforms from a list of callables.
+
+    If sampling up to the max_samples, each invocation of SomeOf uniformly samples a target number of transforms
+    for that invocation.
+
+    At each invocation the set of transforms to apply is uniformly sampled without replacement.
+
+    Args:
+        transforms: list of callables.
+        map_items: whether to apply transform to each item in the input `data` if `data` is a list or tuple.
+            Defaults to `True`.
+        unpack_items: whether to unpack input `data` with `*` as parameters for the callable function of transform.
+            Defaults to `False`.
+        log_stats: whether to log the detailed information of data and applied transform when error happened,
+            for NumPy array and PyTorch Tensor, log the data shape and value range,
+            for other metadata, log the values directly. Default to `False`.
+        min_samples: minimum number of transforms to sample. Defaults to `0`.
+        max_samples: maximum number of transforms to sample. Defaults to `3`.
+        fixed: whether to sample exactly `max_samples` transforms, or up to it. Defaults to `False`.
+    """
+
+    def __init__(
+        self,
+        transforms: Sequence[Callable] | Callable | None = None,
+        map_items: bool = True,
+        unpack_items: bool = False,
+        log_stats: bool = False,
+        min_samples: int = 0,
+        max_samples: int = 3,
+        fixed: bool = True,
+    ) -> None:
+        super().__init__(transforms, map_items, unpack_items, log_stats)
+        if transforms is None:
+            transforms = []
+        self.original_transforms = transforms
+        self.n_transforms = len(transforms)
+        self.max_samples = min(self.n_transforms, max_samples)
+        self.fixed = fixed
+        self.min_samples = min_samples
+        self.random = random.Random(get_seed())
+
+    def __call__(self, data):
+        if self.n_transforms != 0:
+            sample_size = self.max_samples if self.fixed else self.random.randint(self.min_samples, self.max_samples)
+            self.active = self.random.sample(np.arange(self.n_transforms).tolist(), sample_size)
+            self.transforms = [self.original_transforms[i] for i in self.active]
+
+        result = super().__call__(data)
+        return result
```

Comment:
Could you please help also add the `inverse` function in this class?

Reply:
Excuse my ignorance of the code base, but since `SomeOf` inherits `Compose` class, and `Compose` class implements an `inverse` function, what is the benefit here of specifically implementing another `inverse` function inside of `SomeOf`, and how should it differ from the implementation in `Compose`?

Reply:
Ok, so I've taken a look at your implementation and how it works. Technically speaking, it looks like you don't need to implement an ```inverse```, but there is a catch. Your implementation modifies the list of transforms destructively on a call by call basis. This is different to how ```OneOf``` and ```RandomOrder``` work, which are the transforms conceptually very close to ```SomeOf```, which leave the transform lists they are initialised with static and instead use metadata to track what transforms were called or what order they were called in. The inverses for these then use the metadata to call the inverses in the right order.

The issue that arises is that any non-trivial use of compose is very likely to cause major issues, from two causes. _Firstly, if you cause ```Compose.__call__``` multiple times before calling inverse, you will and up with the wrong set of transforms when you call the inverse. Secondly, there are other parts of the codebase that interact with the list of transforms in a compose instance, and none of these expect that list to change on a call by call basis._
 
I think it will help to take a look at ```OneOf``` / ```RandomOrder``` and how it does its inverse. ```SomeOf``` and ```OneOf``` are very close conceptually, and I think if you take the same approach that ```OneOf``` does with its metadata, you can implement inverse in the same way that ```OneOf``` does.

```OneOf``` writes the index of the transform that got called into the metadata during ```__call__```. In inverse, it looks at the metadata to figure out which of its transforms was called and then calls inverse on that. Your ```SomeOf``` could write a list of indices for the methods that were called in the forward pass, and then go through that list of indices in reverse. Really the only difference is iterating over a list of indices rather than a single index.